File: src/web/js/views/attack.js
import { API } from '/static/js/api.js';
import { esc } from '/static/js/helpers.js';

/* ═══════════ Page: ATT&CK (#/attack) ═══════════
 * MITRE ATT&CK posture over HSE's versioned core::attack layer — the same data
 * `hse attack {status,coverage,gaps,navigator}` prints, fetched from
 * /api/v1/attack (which computes nothing of its own). HSE is a passive,
 * authorised OSINT collector, so it honestly claims ONE tactic — Reconnaissance
 * (TA0043). Every covered technique lists the registered modules that are its
 * evidence; the gaps are the honest uncovered slice. This is collection reach,
 * not detection effectiveness, and no decorative score is shown. */

function moduleCell(mods){
  if (mods && mods.length) return mods.map(m=>`<code>${esc(m)}</code>`).join(' ');
  return '<span class="text-muted">— entity/relation mapping</span>';
}

export async function renderAttack(v){
  const d = await API.attack();
  const covered = d.covered || [], gaps = d.gaps || [];
  const pct = Math.round((d.coverage_fraction||0)*1000)/10;

  v.innerHTML = `
    <h2>MITRE ATT&amp;CK &nbsp;<small class="text-muted">Enterprise v${esc(d.attack_version)} · ${esc(d.tactic_id)} ${esc(d.tactic_name)}</small>
      <div class="pull-right">
        <a class="btn btn-default btn-sm" href="/api/v1/attack/navigator" download="hse-attack-navigator.json"
           title="Import into the official ATT&amp;CK Navigator"><i class="glyphicon glyphicon-download-alt"></i>&nbsp;Navigator layer</a>
        <button class="btn btn-default btn-sm" onclick="render()"><i class="glyphicon glyphicon-refresh"></i>&nbsp;Refresh</button>
      </div>
    </h2>
    <hr style="margin:8px 0 14px 0">

    <div class="row">
      <div class="col-sm-3"><div class="stat-card"><div class="lab">Tactic in scope</div><div class="val" style="font-size:16px">${esc(d.tactic_name)}</div><div class="text-muted" style="font-size:10px">the one tactic HSE performs collection for</div></div></div>
      <div class="col-sm-3"><div class="stat-card"><div class="lab">Techniques covered</div><div class="val">${d.techniques_covered}/${d.techniques_total}</div></div></div>
      <div class="col-sm-3"><div class="stat-card"><div class="lab">Coverage</div><div class="val">${pct}%</div><div class="text-muted" style="font-size:10px">derived from real module capability</div></div></div>
      <div class="col-sm-3"><div class="stat-card"><div class="lab">Honest gaps</div><div class="val" style="color:${gaps.length?'#8a6d3b':'#3c763d'}">${gaps.length}</div></div></div>
    </div>

    <div class="panel panel-default">
      <div class="panel-heading">Reconnaissance coverage <small class="text-muted">— each technique and the modules that are its evidence</small></div>
      <table class="table table-condensed table-hover" style="margin:0">
        <thead><tr><th style="width:110px">Technique</th><th style="width:260px">Name</th><th>Modules (evidence)</th></tr></thead>
        <tbody>${covered.map(c=>`<tr><td><code>${esc(c.id)}</code></td><td>${esc(c.name)}</td><td style="font-size:11px">${moduleCell(c.modules)}</td></tr>`).join('')}</tbody>
      </table>
    </div>

    <div class="panel panel-default">
      <div class="panel-heading">Coverage gaps <small class="text-muted">— catalogued Reconnaissance techniques no module collects for</small></div>
      ${gaps.length
        ? `<table class="table table-condensed" style="margin:0"><thead><tr><th style="width:110px">Technique</th><th>Name</th></tr></thead><tbody>${gaps.map(g=>`<tr><td><code>${esc(g.id)}</code></td><td>${esc(g.name)}</td></tr>`).join('')}</tbody></table>`
        : '<div class="panel-body text-muted">No gaps: every catalogued Reconnaissance technique is covered.</div>'}
    </div>

    <p class="text-muted" style="font-size:11px">ATT&amp;CK coverage ≠ detection effectiveness: this reports collection reach for TA0043 only. No other tactic is claimed and no decorative score is emitted.</p>
  `;
}

File: src/cli/attack.rs
//! `hse attack <verb>` — MITRE ATT&CK views over HSE's existing, versioned
//! [`core::attack`](crate::core::attack) layer (Enterprise v17.1).
//!
//! HSE is a passive, authorised OSINT collector, so it honestly claims coverage
//! of exactly ONE tactic — Reconnaissance (TA0043). Every "covered" technique
//! here resolves to the registered modules that actually collect for it (its
//! evidence), and the gaps are the honest uncovered slice. No decorative score
//! is emitted: the coverage fraction is derived from real module capability, and
//! `ATT&CK COVERAGE ≠ DETECTION EFFECTIVENESS` — this reports collection reach,
//! not detection.
//!
//! Data authority: [`crate::modules::reconnaissance_coverage`] (registry-wide
//! coverage) and [`crate::modules::technique_module_index`] (technique →
//! evidence modules). This surface only renders them; it computes no coverage of
//! its own.

use clap::Subcommand;

use crate::core::attack::{self, navigator_layer};
use crate::core::error::{Error, Result};
use crate::modules::{reconnaissance_coverage, technique_module_index};

/// The `hse attack` sub-grammar. `--json` gives the machine-readable shape;
/// `navigator` always emits JSON (a Navigator layer file).
#[derive(Subcommand)]
pub enum AttackAction {
    /// ATT&CK posture summary: catalogue version, the tactic in scope, and how
    /// much of it HSE's registry covers.
    Status {
        /// Output as JSON.
        #[arg(long)]
        json: bool,
    },
    /// Reconnaissance (TA0043) coverage: each technique HSE collects for and the
    /// registered modules that are its evidence.
    Coverage {
        /// Output as JSON.
        #[arg(long)]
        json: bool,
    },
    /// Catalogued Reconnaissance techniques NO registered module collects for —
    /// the honest coverage gaps.
    Gaps {
        /// Output as JSON.
        #[arg(long)]
        json: bool,
    },
    /// Emit a MITRE ATT&CK Navigator layer (JSON) for HSE's static coverage,
    /// importable at the ATT&CK Navigator.
    Navigator,
}

/// Dispatch an `hse attack` verb.
pub(super) fn cmd_attack(action: AttackAction) -> Result<()> {
    match action {
        AttackAction::Status { json } => status(json),
        AttackAction::Coverage { json } => coverage(json),
        AttackAction::Gaps { json } => gaps(json),
        AttackAction::Navigator => navigator(),
    }
}

/// Emit a JSON value pretty-printed, or a serialise error.
fn print_json(value: &serde_json::Value) -> Result<()> {
    println!(
        "{}",
        serde_json::to_string_pretty(value)
            .map_err(|e| Error::Other(format!("serialise attack output: {e}")))?
    );
    Ok(())
}

/// The registered modules that implement a technique, as a compact string. A
/// technique covered via an entity/relation-kind mapping rather than a dedicated
/// module is shown as such rather than as a false module claim.
fn modules_for(id: &str) -> String {
    match technique_module_index().get(id) {
        Some(mods) if !mods.is_empty() => mods.join(", "),
        _ => "— (entity/relation mapping)".to_string(),
    }
}

/// `hse attack status` — posture summary.
fn status(json: bool) -> Result<()> {
    let cov = reconnaissance_coverage();
    let total = cov.covered.len() + cov.uncovered.len();
    let modules_mapped = technique_module_index().len();

    if json {
        return print_json(&serde_json::json!({
            "attack_version": attack::ATTACK_VERSION,
            "tactic_id": cov.tactic_id,
            "tactic_name": cov.tactic_name,
            "techniques_total": total,
            "techniques_covered": cov.covered.len(),
            "coverage_fraction": cov.coverage_fraction,
            "techniques_with_module": modules_mapped,
        }));
    }

    println!("MITRE ATT&CK posture — HSE defensive-intelligence layer\n");
    println!(
        "Catalogue version : ATT&CK Enterprise v{}",
        attack::ATTACK_VERSION
    );
    println!(
        "Tactic in scope   : {} {} (the one tactic HSE performs collection for)",
        cov.tactic_id, cov.tactic_name
    );
    println!(
        "Recon coverage    : {}/{} techniques covered ({:.1}%)",
        cov.covered.len(),
        total,
        cov.coverage_fraction * 100.0
    );
    println!("Techniques w/ >=1 module : {modules_mapped}");
    println!(
        "\nCoverage is derived from real module capability, scoped to Reconnaissance \
         only — no other tactic is claimed, and no decorative score is emitted. \
         This reports collection reach, not detection effectiveness."
    );
    Ok(())
}

/// `hse attack coverage` — covered techniques and their evidence modules.
fn coverage(json: bool) -> Result<()> {
    let cov = reconnaissance_coverage();

    if json {
        let covered: Vec<serde_json::Value> = cov
            .covered
            .iter()
            .map(|c| {
                serde_json::json!({
                    "id": c.technique.id,
                    "name": c.technique.name,
                    "modules": technique_module_index().get(c.technique.id),
                })
            })
            .collect();
        return print_json(&serde_json::json!({
            "tactic_id": cov.tactic_id,
            "tactic_name": cov.tactic_name,
            "coverage_fraction": cov.coverage_fraction,
            "covered": covered,
        }));
    }

    println!("MITRE ATT&CK — HSE Reconnaissance (TA0043) coverage\n");
    if cov.covered.is_empty() {
        println!("No techniques covered.");
        return Ok(());
    }
    println!("{:<12}  {:<34}  MODULES (evidence)", "TECHNIQUE", "NAME");
    println!("{}", "─".repeat(96));
    for c in &cov.covered {
        println!(
            "{:<12}  {:<34}  {}",
            c.technique.id,
            c.technique.name,
            modules_for(c.technique.id)
        );
    }
    println!(
        "\n{}/{} techniques covered ({:.1}%).",
        cov.covered.len(),
        cov.covered.len() + cov.uncovered.len(),
        cov.coverage_fraction * 100.0
    );
    Ok(())
}

/// `hse attack gaps` — catalogued Reconnaissance techniques no module reaches.
fn gaps(json: bool) -> Result<()> {
    let cov = reconnaissance_coverage();

    if json {
        let gaps: Vec<serde_json::Value> = cov
            .uncovered
            .iter()
            .map(|t| serde_json::json!({ "id": t.id, "name": t.name }))
            .collect();
        return print_json(&serde_json::json!({ "gaps": gaps }));
    }

    println!("MITRE ATT&CK — HSE Reconnaissance (TA0043) coverage gaps\n");
    if cov.uncovered.is_empty() {
        println!("No gaps: every catalogued Reconnaissance technique is covered.");
        return Ok(());
    }
    println!("{:<12}  NAME", "TECHNIQUE");
    println!("{}", "─".repeat(60));
    for t in &cov.uncovered {
        println!("{:<12}  {}", t.id, t.name);
    }
    println!(
        "\n{} of {} Reconnaissance techniques are not collected for (honest gaps).",
        cov.uncovered.len(),
        cov.covered.len() + cov.uncovered.len()
    );
    Ok(())
}

/// `hse attack navigator` — a Navigator layer (JSON) for HSE's static coverage.
fn navigator() -> Result<()> {
    let cov = reconnaissance_coverage();
    let layer = navigator_layer(&cov, "HSE static Reconnaissance coverage");
    print_json(&layer)
}
